homeassistant/components/sleepiq/number.py

```diff
@@ -1,15 +1,98 @@
 """Support for SleepIQ SleepNumber firmness number entities."""
-from asyncsleepiq import SleepIQBed, SleepIQSleeper
+from __future__ import annotations
 
-from homeassistant.components.number import NumberEntity
+from collections.abc import Callable, Coroutine
+from dataclasses import dataclass
+from typing import Any, cast
+
+from asyncsleepiq import SleepIQActuator, SleepIQBed, SleepIQSleeper
+
+from homeassistant.components.number import NumberEntity, NumberEntityDescription
 from homeassistant.config_entries import ConfigEntry
 from homeassistant.core import HomeAssistant, callback
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
 
-from .const import DOMAIN, FIRMNESS
+from .const import ACTUATOR, DOMAIN, ENTITY_TYPES, FIRMNESS, ICON_OCCUPIED
 from .coordinator import SleepIQData
-from .entity import SleepIQSleeperEntity
+from .entity import SleepIQBedEntity
+
+
+@dataclass
+class SleepIQNumberEntityDescriptionMixin:
+    """Mixin to describe a SleepIQ number entity."""
+
+    value_fn: Callable[[Any], float]
+    set_value_fn: Callable[[Any, int], Coroutine[None, None, None]]
+    get_name_fn: Callable[[SleepIQBed, Any], str]
+    get_unique_id_fn: Callable[[SleepIQBed, Any], str]
+
+
+@dataclass
+class SleepIQNumberEntityDescription(
+    NumberEntityDescription, SleepIQNumberEntityDescriptionMixin
+):
+    """Class to describe a SleepIQ number entity."""
+
+
+async def _async_set_firmness(sleeper: SleepIQSleeper, firmness: int) -> None:
+    await sleeper.set_sleepnumber(firmness)
+
+
+async def _async_set_actuator_position(
+    actuator: SleepIQActuator, position: int
+) -> None:
+    await actuator.set_position(position)
+
+
+def _get_actuator_name(bed: SleepIQBed, actuator: SleepIQActuator) -> str:
+    if actuator.side:
+        return f"SleepNumber {bed.name} {actuator.side_full} {actuator.actuator_full} {ENTITY_TYPES[ACTUATOR]}"
+
+    return f"SleepNumber {bed.name} {actuator.actuator_full} {ENTITY_TYPES[ACTUATOR]}"
+
+
+def _get_actuator_unique_id(bed: SleepIQBed, actuator: SleepIQActuator) -> str:
+    if actuator.side:
+        return f"{bed.id}_{actuator.side}_{actuator.actuator}"
+
+    return f"{bed.id}_{actuator.actuator}"
+
+
+def _get_sleeper_name(bed: SleepIQBed, sleeper: SleepIQSleeper) -> str:
+    return f"SleepNumber {bed.name} {sleeper.name} {ENTITY_TYPES[FIRMNESS]}"
+
+
+def _get_sleeper_unique_id(bed: SleepIQBed, sleeper: SleepIQSleeper) -> str:
+    return f"{sleeper.sleeper_id}_{FIRMNESS}"
```

Comment:
Do we need the `bed.id` here or is `sleeper_id` globally unique?

Reply:
The `sleeper_id` is in the same format as the `bed_id` (19 digits preceded by a hyphen) so I'm pretty sure it's globally unique. The main reason I wanted to remove `bed_id` is so that it's easier to detect when a unique id needs to be migrated